File: src/main/java/com/thetaris/goldprice/service/GoldPriceService.java
package com.thetaris.goldprice.service;

import com.thetaris.goldprice.WsdlProperties;
import com.thetaris.goldprice.generated.GetCurrentGoldPrice;
import com.thetaris.goldprice.generated.GetCurrentGoldPriceResponse;
import org.springframework.oxm.jaxb.Jaxb2Marshaller;
import org.springframework.stereotype.Service;
import org.springframework.ws.client.core.WebServiceTemplate;
import org.springframework.ws.soap.client.core.SoapActionCallback;

@Service
public class GoldPriceService {

    private final WebServiceTemplate webServiceTemplate;

    private final GetCurrentGoldPrice getCurrentGoldPrice;

    private final WsdlProperties wsdlProperties;

    public GoldPriceService(GetCurrentGoldPrice getCurrentGoldPrice, WsdlProperties wsdlProperties, Jaxb2Marshaller marshaller) {
        this.getCurrentGoldPrice = getCurrentGoldPrice;
        this.wsdlProperties = wsdlProperties;

        webServiceTemplate = new WebServiceTemplate();
        webServiceTemplate.setMarshaller(marshaller);
        webServiceTemplate.setUnmarshaller(marshaller);
    }

    public GetCurrentGoldPriceResponse getPrices() {
        return (GetCurrentGoldPriceResponse) webServiceTemplate
                .marshalSendAndReceive(
                        wsdlProperties.getUrl(),
                        getCurrentGoldPrice,
                        new SoapActionCallback(wsdlProperties.getAction()));
    }

}

File: src/main/kotlin/com/thetaris/goldprice/service/GoldPriceService.kt
package com.thetaris.goldprice.service

import com.thetaris.goldprice.config.WsdlProperties
import com.thetaris.goldprice.generated.GetCurrentGoldPrice
import com.thetaris.goldprice.generated.GetCurrentGoldPriceResponse
import org.springframework.oxm.jaxb.Jaxb2Marshaller
import org.springframework.stereotype.Service
import org.springframework.ws.client.core.WebServiceTemplate
import org.springframework.ws.soap.client.core.SoapActionCallback

@Service
class GoldPriceService(
        private val getCurrentGoldPrice: GetCurrentGoldPrice,
        private val wsdlProperties: WsdlProperties,
        marshaller: Jaxb2Marshaller
) {

    private final val webServiceTemplate: WebServiceTemplate = WebServiceTemplate()

    init {
        webServiceTemplate.marshaller = marshaller
        webServiceTemplate.unmarshaller = marshaller
    }

    fun getPrices(): GetCurrentGoldPriceResponse =
            webServiceTemplate.marshalSendAndReceive(wsdlProperties.url, getCurrentGoldPrice, getAction())
                    as GetCurrentGoldPriceResponse

    private fun getAction() = SoapActionCallback(wsdlProperties.action)

}
